docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,268 @@
+
+The following limits exist for Azure Compute Galleries:
+1. 100 galleries, per subscription, per region
+1. 1,000 image definitions, per subscription, per region
+1. 10,000 image versions, per subscription, per region
+1. 100 replicas per image version however 50 replicas should be sufficient for most use cases
+1. Any disk attached to the image must be less than or equal to 1 TB in size
+1. Resource move isn't supported for Azure compute gallery resources
+
+Link to page with Azure Compute Gallery limits: https://learn.microsoft.com/en-us/azure/virtual-machines/azure-compute-gallery
+
+Thus the feature will not work for OS images with disks attached greater than 1 TB in size. These limits should be kept in mind by the operator since this feature requires the use of an Azure Compute Gallery. The Azure Compute Gallery itself costs no money. For each image, it costs roughly $0.10 - $0.13 for a CAPI image per region and $0.1 - $0.3 for a plain Flatcar image per region with the Standard HDD LRS storage account type every day. Scaling across other regions will also accrue network egress charges. Thus, the feature may cost the operator more money if they don't need faster horizontal scaling.
+
+The UX will mostly be impactful towards operators and members of the CAPZ community will test these changes and give feedback on them. Security will also likely follow in terms of how it gets reviewed, but no major security problems should be possible from this change. For folks who work outside the SIG or subproject, they should hopefully have faster horizontal scaling without needing to directly do anything outside of setting an environment variable on clusterctl initialization and updating their AzureMachinePools.
+
+## Alternatives
+
+No currently known alternatives exist which are public and have been implemented for CAPZ. A savvy operator may also have created a similar system for their nodes but we hope to offload that responsibility to the CAPZ team in this process, and currently no other open source implementations of this for CAPZ are known to exist. 
+
+In terms of how we choose a node as the prototype node, lots of different metrics or heuristics can be used like manually creating a temporary prototype for testing, having the fastest ready time, or anything else which is seen as typically good but a more generalist approach is outlined here since more specific methods may not be as helpful for certain operators.
+
+For architectural details of where else the code could exist, the controller section makes the most sense since this proposal will be constantly modifying the state of our objects, but theoretically it could be largely put into hack with shell scripts and then a controller could simply be ordered to trigger that shell script, but this is less maintainable in the long run and not as preferred. 
```

Comment:
One other related aspect I'm not sure is discussed in the doc is putting this code in the AzureMachinePool controller vs. creating a new controller.

Reply:
Added a sentence here about this tradeoff/discussion